test/e2e/federated-secret.go

```diff
@@ -18,69 +18,166 @@ package e2e
 			secret := createSecretOrFail(f.FederationClientset_1_4, nsName)
-			By(fmt.Sprintf("Creation of secret %q in namespace %q succeeded.  Deleting secret.", secret.Name, nsName))
-			// Cleanup
-			err := f.FederationClientset_1_4.Core().Secrets(nsName).Delete(secret.Name, &api.DeleteOptions{})
-			framework.ExpectNoError(err, "Error deleting secret %q in namespace %q", secret.Name, secret.Namespace)
-			By(fmt.Sprintf("Deletion of secret %q in namespace %q succeeded.", secret.Name, nsName))
-		})
 
+			defer func() { // Cleanup
+				By(fmt.Sprintf("Deleting secret %q in namespace %q", secret.Name, nsName))
+				err := f.FederationClientset_1_4.Core().Secrets(nsName).Delete(secret.Name, &api.DeleteOptions{})
+				framework.ExpectNoError(err, "Error deleting secret %q in namespace %q", secret.Name, nsName)
+			}()
+			// wait for secret shards being created
+			waitForSecretShardsOrFail(nsName, secret, clusters)
+			secret = updateSecretOrFail(f.FederationClientset_1_4, nsName)
+			waitForSecretShardsUpdatedOrFail(nsName, secret, clusters)
+		})
 	})
 })
 
 func createSecretOrFail(clientset *federation_release_1_4.Clientset, namespace string) *v1.Secret {
 	if clientset == nil || len(namespace) == 0 {
-		Fail(fmt.Sprintf("Internal error: invalid parameters passed to deleteSecretOrFail: clientset: %v, namespace: %v", clientset, namespace))
+		Fail(fmt.Sprintf("Internal error: invalid parameters passed to createSecretOrFail: clientset: %v, namespace: %v", clientset, namespace))
 	}
-	By(fmt.Sprintf("Creating federated secret %q in namespace %q", FederatedSecretName, namespace))
 
 	secret := &v1.Secret{
 		ObjectMeta: v1.ObjectMeta{
 			Name: FederatedSecretName,
 		},
 	}
-
-	By(fmt.Sprintf("Trying to create secret %q in namespace %q", secret.Name, namespace))
+	By(fmt.Sprintf("Creating secret %q in namespace %q", secret.Name, namespace))
 	_, err := clientset.Core().Secrets(namespace).Create(secret)
-	framework.ExpectNoError(err, "Creating secret %q in namespace %q", secret.Name, namespace)
+	framework.ExpectNoError(err, "Failed to create secret %s", secret.Name)
 	By(fmt.Sprintf("Successfully created federated secret %q in namespace %q", FederatedSecretName, namespace))
 	return secret
 }
+
+func updateSecretOrFail(clientset *federation_release_1_4.Clientset, namespace string) *v1.Secret {
+	if clientset == nil || len(namespace) == 0 {
+		Fail(fmt.Sprintf("Internal error: invalid parameters passed to updateSecretOrFail: clientset: %v, namespace: %v", clientset, namespace))
+	}
+
+	var err error
+	var newSecret *v1.Secret
+	secret := &v1.Secret{
+		ObjectMeta: v1.ObjectMeta{
+			Name: UpdatedFederatedSecretName,
+		},
+	}
+
+	for retryCount := 0; retryCount < MaxRetries; retryCount++ {
+		_, err = clientset.Core().Secrets(namespace).Get(FederatedSecretName)
+		if err != nil {
+			framework.Failf("failed to get secret %q: %v", FederatedSecretName, err)
+		}
+		newSecret, err = clientset.Core().Secrets(namespace).Update(secret)
```

Comment:
This command is failing with:
```
failed to update secret "federated-secret": secrets "updated-federated-secret" not found
```

We cannot change the name of a resource while updating it.

cc @kshafiee